### PR TITLE
batch_normalization: Introduce vectorization optimization in the batch norm elementwise kernel.

### PR DESCRIPTION
**Pull request by xytintel**
Due to performance issues with the low-precision data type implementation of group stride loops on PVC ([jira: PYTORCHDGQ-5162](https://jira.devtools.intel.com/browse/PYTORCHDGQ-5162)), partial vectorization optimization is used.

### CODE REVIEW

File: src/ATen/native/xpu/sycl/BatchNormKernels.cpp [hunk 1]
```diff
@@ -1263,6 +1263,147 @@ struct BatchNormTransformInputKernelFunctor {
...
+                gamma * (i[feature] - mean) * invstd + beta);
+          }
+        } else {
+          using vec_t = memory::aligned_vector<input_scalar_t, VEC_SIZE>;
+          vec_t vec;
+#pragma unroll
+          for (int vt = 0; vt < VEC_SIZE; ++vt) {
+            index_t feature = feature_vec_begin + vt;
+            vec[vt] = static_cast<input_scalar_t>(
+                gamma * (i[feature] - mean) * invstd + beta);
```
Comment (fengyuan14) [inline src/ATen/native/xpu/sycl/BatchNormKernels.cpp:1323]: Why not using vectorization for inputs? I see `vt` is contiguous across the iteration.

File: src/ATen/native/xpu/sycl/BatchNormKernels.cpp [hunk 2]
```diff
@@ -1319,14 +1460,27 @@ void batch_norm_elemt_template(
...
+      memory::can_vectorize_up_to<input_scalar_t>(output_ptr) >= 4 &&
+      sizeof(input_scalar_t) < sizeof(float)) {
+    auto kfn = BatchNormTransformInputVectorizedKernelFunctor<
+        4,
+        input_scalar_t,
+        stat_scalar_t,
+        stat_accscalar_t,
+        true,
+        index_t>(input, output, mean, invstd, weight, bias, dummy_epsilon);
+    sycl_kernel_submit(global_range, local_range, queue, kfn);
```
Comment (fengyuan14) [inline src/ATen/native/xpu/sycl/BatchNormKernels.cpp:1474]: Why we still keep the `global_range` unchanged, even the workloads on each work items increase?

File: src/ATen/native/xpu/sycl/BatchNormKernels.cpp (hunk 1 above)
Comment (fengyuan14) [reply src/ATen/native/xpu/sycl/BatchNormKernels.cpp:1325]: If we know the `vt` size, we can calculate offset with linear id 0, 3, 7, 11? if the `vt` size is 4?

**Comment by EikanWang** [comment]
@xytintel , may I know if this PR is still available?

**Comment by xytintel** [comment]
> @xytintel , may I know if this PR is still available?

Yes, In the process of merging it.

**Comment by xytintel** [comment]
Group stride kernel: 12.493ms
Vectorized kernel: 10.024ms
```python
import torch
a = torch.rand(2, 256, 128, 128).xpu().bfloat16()
bn = torch.nn.BatchNorm2d(256).xpu().bfloat16()
prof_xpu = torch.profiler.profile(
    activities=[
    torch.profiler.ProfilerActivity.CPU,
    torch.profiler.ProfilerActivity.XPU],
)
with prof_xpu:
    for i in range(120):
        output = bn(a)
print(prof_xpu.key_averages().table(sort_by="self_xpu_time_total", row_limit=100000))
print(output.dtype)
```

File: src/ATen/native/xpu/sycl/BatchNormKernels.cpp (hunk 1 above)
Comment (xytintel) [reply src/ATen/native/xpu/sycl/BatchNormKernels.cpp:1319]: Apply a coalescing manner for the input access would be quite complex, let we discuss it later.

**Review by EikanWang — APPROVED** [review]
